### PR TITLE
Refactor and Enhance Agent Protocol

### PR DESCRIPTION
**Pull request by Swiftyos**
## Description

We are excited to announce a series of enhancements and refactoring in our Agent Protocol implementation. These changes improve flexibility, extensibility, and integration options. The following key changes have been introduced:

## Changes:

- **Refactored Agent Protocol for Subclassing Compatibility**: The agent protocol has been refactored to support subclassing, ensuring compatibility when the protocol is extended. This enhancement empowers developers to customize the agent's behavior to meet specific requirements.
- **Using Middleware to  Inject the Agent into endpoints**: The process of integrating the agent with FastAPI has been streamlined using more efficient middleware injection.
- **Simplified Handler functionality**: Switched to calling functions from the injected agent
- **Removed Opinionated Artifact handling**: Artifact handling can now be implemented by the Agent developer

### CODE REVIEW

**Comment by Swiftyos** [comment]
```
python git:(main) ✗ agent-protocol test --url http://127.0.0.1:8000 
===================================================== test session starts ======================================================
platform linux -- Python 3.11.2, pytest-7.4.0, pluggy-1.2.0 -- /home/arty/.cache/pypoetry/virtualenvs/agent-protocol-hR6WbidJ-py3.11/bin/python
cachedir: .pytest_cache
rootdir: /home/arty/dev/ai/agent-protocol/sdk/python
plugins: anyio-3.7.1
collected 7 items                                                                                                              

agent_protocol/utils/compliance/main.py::TestCompliance::test_create_agent_task[http://127.0.0.1:8000] PASSED            [ 14%]
agent_protocol/utils/compliance/main.py::TestCompliance::test_list_agent_tasks_ids[http://127.0.0.1:8000] PASSED         [ 28%]
agent_protocol/utils/compliance/main.py::TestCompliance::test_get_agent_task[http://127.0.0.1:8000] PASSED               [ 42%]
agent_protocol/utils/compliance/main.py::TestCompliance::test_list_agent_task_steps[http://127.0.0.1:8000] PASSED        [ 57%]
agent_protocol/utils/compliance/main.py::TestCompliance::test_execute_agent_task_step[http://127.0.0.1:8000] PASSED      [ 71%]
agent_protocol/utils/compliance/main.py::TestCompliance::test_list_artifacts[http://127.0.0.1:8000] PASSED               [ 85%]
agent_protocol/utils/compliance/main.py::TestCompliance::test_get_agent_task_step[http://127.0.0.1:8000] PASSED          [100%]

====================================================== 7 passed in 0.05s =======================================================
```

File: sdk/python/agent_protocol/middlewares.py [hunk 1]
```diff
@@ -11,3 +11,35 @@ async def not_found_exception_handler(
...
+
+        Examples:
+            >>> from fastapi import FastAPI, Request
+            >>> from agent_protocol.agent import Agent
+            >>> from agent_protocol.middlewares import AgentMiddleware
+            >>> app = FastAPI()
+            >>> @app.get("/")
+            >>> async def root(request: Request):
+            >>>     agent = request["agent"]
+            >>>     return agent
```
Comment (jakubno) [inline sdk/python/agent_protocol/middlewares.py:36]: Isn't this confusing? Why are you returning agent?

File: sdk/python/examples/minimal.py [hunk 2]
```diff
@@ -1,16 +1,36 @@
-from agent_protocol import Agent, Task, Step
+from agent_protocol import Agent, Artifact, Step, Task
+from agent_protocol.db import InMemoryTaskDB
 
 
-async def task_handler(task: Task) -> None:
-    print(f"task: {task.input}")
-    await Agent.db.create_step(task.task_id, task.input)
+class AutoGPT(Agent):
```
Comment (jakubno) [inline sdk/python/examples/minimal.py:5]:
```suggestion
class MinimalAgent(Agent):
```

File: openapi.yml [hunk 3]
```diff
@@ -248,20 +248,32 @@ components:
...
-          example: 'main.py'
-        relative_path:
-          description: Relative path of the artifact in the agent's workspace.
+          description: URI of the artifact.
+          example: file://home/bob/workspace/bucket/main.py
+        data:
           type: string
-          example: 'python/code/'
+          format: byte
+          description: Data of the artifact.
```
[the remark this answers is not in the document]
Comment (Swiftyos) [reply openapi.yml:276]: I agree. This wasn't a wise addition. Will revert

File: sdk/python/agent_protocol/middlewares.py (hunk 1 above)
Comment (Swiftyos) [reply sdk/python/agent_protocol/middlewares.py:36]: The returning the agent in the example is confusing will update

File: openapi.yml [hunk 4]
```diff
@@ -248,20 +248,32 @@ components:
...
           description: Filename of the artifact.
+          example: main.py
+        saved:
+          type: boolean
+          description: Whether the artifact has been saved to the agent's workspace.
+          example: false
+        agent_created:
+          type: boolean
+          description: Whether the artifact has been created by the agent.
+          example: false
```
[the remark this answers is not in the document]
Comment (Swiftyos) [reply openapi.yml:268]: My thinking was I wanted a way to know if the Artefact had been successfully saved. So we would know to return an error if we had been unable to save.

However, I don't the solution to that problem is adding a saved bool to the artefact. - Will remove

I do feel `agent_created` is a useful concept

**Comment by Swiftyos** [comment]
Added the heartbeat change here as well. As it seems to have been forgotten about

**Comment by Swiftyos** [comment]
Closing PR as this modifies the SDK rather than being focused on the protocol

**Comment by jakubno** [comment]
The openapi spec has to be changed first. Unfortunately this wasn't following the protocol

**Pull request closed without merging**